### PR TITLE
[share_plus] Enable Web implementations for `shareFiles`

### LINKED ISSUE #212: Missing share_plus share files implementation in web
There is a share files method in the html.navigator, see [here](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/share)

### PR DESCRIPTION
**Pull request by ThexXTURBOXx**
## Description

This PR is a continuation of #318.
There were a few issues in that original implementation:
 - The int arrays were passed as a concatenated int string to the sharing program
 - File name being forgotten
 - Mime Types missing (causing `Permission Denied` errors)
 - The `withResult` implementations were missing
 - The examples were not updated properly
 - Tests were missing

This PR aims at fixing these issues as they seemed to be the remaining ones.
Also, I have tested the feature on Chrome (Android + PC) and Edge (PC) and it worked in my test cases (multiple files, different mime types etc.).
Firefox is sadly not supported (due to Browser limitations, see [here](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/share#browser_compatibility) for details).

## Related Issues

Fixes #212

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`).
This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated the version in `pubspec.yaml` and `CHANGELOG.md`.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/fluttercommunity/plus_plugins/blob/main/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning

### CODE REVIEW

**Comment by ThexXTURBOXx** [comment]
Sorry for the odd title at first. Something on my end appearently went wrong there. Fixed it, though!

**Comment by daniel-shuy** [comment]
The `share_plus` example currently uses [image_picker](https://pub.dev/packages/image_picker), which does not work on Flutter Web. The example should be updated to use [image_picker_web](https://pub.dev/packages/image_picker_web) if the platform is web, so that this feature can be tested with the `share_plus` example using a mobile/emulator browser, by running the example on Web Server.

**Comment by ThexXTURBOXx** [comment]
I have successfully tested the example on web, even the image picker. Why should the image picker not work there?

File: packages/share_plus/share_plus_web/lib/share_plus_web.dart [hunk 1]
```diff
@@ -61,7 +62,107 @@ class SharePlusPlugin extends SharePlatform {
...
+        context: DiagnosticsNode.message('while trying to share file(s)'),
+        library: 'share_plus_web',
+        informationCollector: () => [
+          DiagnosticsNode.message(
+            'The web share API is not available on the current browser. '
+            'Falling back to downloading the files. '
+            'Subject and text are discarded if given.',
+          ),
+        ],
+      ));
```
Comment (miquelbeltran) [inline packages/share_plus/share_plus_web/lib/share_plus_web.dart:118]: Should we call directly into the FlutterError error sink? 

I think it would be better to throw an exception instead and let the developer catch it on their side.

File: packages/share_plus/share_plus_platform_interface/lib/method_channel/method_channel_share.dart [hunk 2]
```diff
@@ -69,6 +70,26 @@ class MethodChannelShare extends SharePlatform {
...
   }
 
+  /// Summons the platform's share sheet to share multiple files.
+  @override
+  Future<void> shareXFiles(
+    List<XFile> files, {
+    String? subject,
+    String? text,
+    Rect? sharePositionOrigin,
+  }) {
```
Comment (miquelbeltran) [inline packages/share_plus/share_plus_platform_interface/lib/method_channel/method_channel_share.dart:80]: Is it necessary to have the `shareXFiles` method in the platform interface, rather than being a private method of the web implementation?

File: packages/share_plus/share_plus/lib/share_plus.dart [hunk 3]
```diff
@@ -82,6 +84,42 @@ class Share {
...
+  /// or handled.
+  /// On iOS image/jpg, image/jpeg and image/png are handled as images, while
+  /// every other MIME type is considered a normal file.
+  ///
+  /// The optional `sharePositionOrigin` parameter can be used to specify a global
+  /// origin rect for the share sheet to popover from on iPads and Macs. It has no effect
+  /// on other devices.
+  ///
+  /// May throw [PlatformException] or [FormatException]
+  /// from [MethodChannel].
```
Comment (miquelbeltran) [inline packages/share_plus/share_plus/lib/share_plus.dart:107]: Could you document here instead what is the difference between `shareFile` and `shareXFile` and when should `shareXFile` be used?

I understand we need `shareXFile` because `XFile` is what web uses, correct?

**Comment by miquelbeltran** [comment]
I am looking into it and added some questions, but overall I wonder if we could avoid adding a new share method for `XFiles`.

Also, don't create two new methods with and without result, but rather just one method with result. The share methods without result should be legacy.

The final goal is to have a unique universal share method as the current implementation is a mess.

**Comment by ThexXTURBOXx** [comment]
@miquelbeltran Thank you very much for the feedback. I will implement it as soon as possible today. I will remove the shareXFiles function without result and deprecate the `shareFiles*` functions since they are obsolete now with the `shareXFiles*` functions.
However, I won't unify all functions into one in this PR since it is hugely out of scope and described in another issue.

**Comment by miquelbeltran** [comment]
Sounds good to me, thanks!

File: packages/share_plus/share_plus/lib/share_plus.dart (hunk 3 above)
Comment (ThexXTURBOXx) [reply packages/share_plus/share_plus/lib/share_plus.dart]: Basically yes, `XFile` is a cross-platform `File` implementation which makes it very handy to use on all platforms.

File: packages/share_plus/share_plus_platform_interface/lib/method_channel/method_channel_share.dart (hunk 2 above)
Comment (ThexXTURBOXx) [reply packages/share_plus/share_plus_platform_interface/lib/method_channel/method_channel_share.dart]: This is also solved because of the deprecations

File: packages/share_plus/share_plus_web/lib/share_plus_web.dart (hunk 1 above)
Comment (ThexXTURBOXx) [reply packages/share_plus/share_plus_web/lib/share_plus_web.dart]: Done, the exceptions are now uncaught

**Review by miquelbeltran — APPROVED** [review]
LGTM! thanks for applying all suggestions
